### ISSUE TITLE
🐛 Sequelize throws SQLITE_BUSY error on parallel gateways

### ISSUE THREAD

**Issue by robpal94** [process-engine/process_engine_runtime#287]
#### Describe your issue

Consider this Process: 
![parallel_lanes_test_bpmn](https://user-images.githubusercontent.com/15343316/53952527-4155df00-40d1-11e9-91b7-346778bad878.png)

You can see, that there are many outgoing sequence flows from a parallel gateway.

When executing this Process, the ProcessEngine throws the following error: 
```
2019-03-07T10:42:17.070Z - error: [processengine:persistence:flow_node_instance_repository] Failed to persist new instance for FlowNode ScriptTask_0rowo17, using instance id ...! { SequelizeTimeoutError: SQLITE_BUSY: database is locked
    at Query.formatError (/Users/robin/Documents/process_engine/sequelize_connection_manager/node_modules/sequelize/lib/dialects/sqlite/query.js:420:16)
  ...
  name: 'SequelizeTimeoutError',
  parent:
   { Error: SQLITE_BUSY: database is locked
     errno: 5,
     code: 'SQLITE_BUSY',
     sql: 'INSERT INTO `FlowNodeInstances` (`id`,`flowNodeInstanceId`,`flowNodeId`,`flowNodeType`,`correlationId`,`processModelId`,`processInstanceId`,`identity`,`state`,`previousFlowNodeInstanceId`,`createdAt`,`updatedAt`) VALUES (NULL,\'f10c6432-6b73-4b04-8aff-e8e0e517e8c4\',\'ScriptTask_0rowo17\',\'bpmn:ScriptTask\',\'3b3d1616-9e5f-4b2c-82de-d281fd1b1552\',\'paralell_process\',\'8a8dd4d3-2c01-4f35-8cef-ca10afbf0d46\',\'{"token":"ZHVtbXlfdG9rZW4=","userId":"dummy_token"}\',\'running\',\'770ceb6b-af53-4b62-a265-a8bb7e875d29\',\'2019-03-07 10:42:07.013 +00:00\',\'2019-03-07 10:42:07.013 +00:00\');' },
```

#### Cause

It seems that this is because SQLite can only handle one transaction at time.
* If one transaction is currently running and another transaction request against the Database is made,
   the new transaction will be transferred into a `waiting` - State. 
* If a transaction lingers for a longer amount in this waiting state, SQLite throws the `Database is locked` exception to prevent deadlocks.

This issue seems to occur because we seem to shoot multiple transactions against the database.
* All these gets enqueued
* Some of them are running into this _linger timeout_
* SQlite throws the `Database is locked` error, which is not handled.

This blogpost kinda explains it a bit better: http://beets.io/blog/sqlite-nightmare.html

#### How can others reproduce the issue?

* Execute this Process: [paralell_process.bpmn.txt](https://github.com/process-engine/process_engine_runtime/files/2940830/paralell_process.bpmn.txt)
* See that the ProcessEngine freaks out

#### Possible solution

The most sloppy solution would be to add a `retry` option. This way, sequelize would simply retry the query n times, if the error occurs. 

This can be done like so:
```JavaScript
retry: {
    match: [
       /SQLITE_BUSY/,
     ],
     name: 'query',
     max: 5
},
```

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

**Comment by ElRaptorus** [comment]
Could you do us all a favour and downscale your images next time? 
I feel like I am looking at some new tapestry for our office.

**Comment by ElRaptorus** [comment]
I took the liberty of downscaling the image.
Next time, please consider using a resoltion that is a little lower than 2400x3600. 😉

**Comment by ElRaptorus** [comment]
Regarding the issue, I cannot say for sure where this is coming from, but my prime suspect would be the transactions we use with the `FlowNodeInstanceRepository`.

[Here](https://github.com/process-engine/flow_node_instance.repository.sequelize/blob/develop/src/flow_node_instance_repository.ts#L447) and [here](https://github.com/process-engine/flow_node_instance.repository.sequelize/blob/develop/src/flow_node_instance_repository.ts#L540).

The transactions are meant to ensure that now FlowNodeInstance will be persisted without a corresponding ProcessToken. They also perform a proper rollback, in case one of the queries fail.

I suspect the problem may be the fact, that the transactions are [unmanaged](http://docs.sequelizejs.com/manual/tutorial/transactions.html#unmanaged-transaction-then-callback-), but I cannot be certain.
I tried to use managed transactions some time ago, but that caused a lot of problems. The queries would keep resolving, before all parts of the transaction were actually finished, which in turn lead to a lot of follow-up requests failing.

Another possible source for this error could be the fact that the repositories run as [singleton](https://github.com/process-engine/flow_node_instance.repository.sequelize/blob/develop/ioc_module.js#L12).
I don't know that any of the repositories are used in more than one location, but it could still be a contributing factor.

I'm afraid you're just going to have to start digging, just like I did.

**Comment by robpal94** [comment]
> Could you do us all a favour and downscale your images next time?
I feel like I am looking at some new tapestry for our office.

I actually didn't paid attention to the size of this image. I just pasted the generated image from BPMN Studio (in github, it looks fine).

**Comment by ElRaptorus** [comment]
> (in github, it looks fine).

Yes, it looks fine *now*, because I scaled it down. 😉

**Comment by robpal94** [comment]
So, I equipped my headlamp and start digging through the SQLite C implementation to make myself a clearer pictures, what causes the `SQLite Busy` exception. 

I found a few Bits, what seems to lead to this. _(Please take this with a grain of salt)_ :

_Note: Even if the SourceFiles a pretty big, they are nicely documented. If you really follow the lines I've linked, you can scroll up and read the documentation comments._ 

The root _exception_ seems to be thrown here: 
https://github.com/mackyle/sqlite/blob/master/src/wal.c#L3004

When trying to create a new transaction, sqlite tries to obtain a new Lock for the database file. If the file is already locked, `SQLITE_BUSY` is _set_.
* This happens if we try to execute a write transaction while the database is currently busy with another write transaction.

If the lock couldn't get obtained, the transaction also couldn't be opened.
Now sqlite tries this again, as found here: https://github.com/mackyle/sqlite/blob/master/src/btree.c#L3454

As you can see, `while` executes the `btreeInvokeBusyHandler()` Function.
* Its possible to implement your own _handler_ Functions when the database is busy
* I don't think that this is exported to the `node-sqlite` adapter but I may be wrong.

A default handler, defined [here](https://github.com/mackyle/sqlite/blob/master/src/main.c#L1513) is called. This handler checks, if the maximum delay was encountered.
* If not, return 1. This leads the above mentioned `while` loop to keep trying to obtain the lock for the transaction.
* If so, return 0. This leads the above mentioned `while` loop to exit.
https://github.com/mackyle/sqlite/blob/master/src/main.c#L1510

#### TL;DR
Error is thrown because too many transactions are opened/the database is too slow; In theory, the busy behavior can be changed by a custom handler; Not sure if this is possible with node-sqlite and sequelize

**Comment by robpal94** [comment]
I also tried to start 50 ProcessInstance of a simple Process, which looks like so:

![simple_process](https://user-images.githubusercontent.com/38498731/54042302-17ccae80-41ca-11e9-98f0-f568fb2267f8.png)

In this case, the `SQLITE_BUSY` error also occurs.

**Comment by ElRaptorus** [comment]
@robpal94 This is some excellent dig-work, thank you.

I think what you did in the Sequelize Connection Manager is the best way to go for now.
